### PR TITLE
fix(comments): keep reply composer above keyboard

### LINKED ISSUE #5939: fix(comments): keep reply composer above keyboard
## Problem

When replying to a comment on a post, the reply composer can slide under the on-screen keyboard and stop being usable. The user can keep typing, but the text and action buttons are hidden behind the keyboard. Once that happens, the visible UI does not provide a reliable way to dismiss the keyboard or tap the send button.

This happens most often in reply mode and can sometimes happen for initial comments.

## Current diagnosis

The comments surface is shown through `showVideoPausingVineBottomSheet`, which delegates to `VineBottomSheet.show` with:

- `isScrollControlled: true`
- `initialChildSize: 0.7`
- `minChildSize: 0.5`
- `maxChildSize: 0.93`
- `snapSizes: [0.7, 0.93]`

The sheet body is a fixed-height `DraggableScrollableSheet`. `VineBottomSheet` already pads the bottom input by `MediaQuery.viewInsets.bottom`, but that padding is applied inside the sheet's fixed fractional height. When the keyboard appears, the fixed children can exceed the available sheet height:

- header
- divider
- reply composer
- reply indicator
- keyboard-height bottom padding

At smaller sheet sizes, the scrollable area collapses and the composer controls can be pushed below the visible region. Reply mode is more vulnerable because the input is multiline and includes the reply indicator row.

Relevant code:

- `mobile/lib/screens/comments/comments_screen.dart`
- `mobile/lib/screens/comments/widgets/comment_input.dart`
- `mobile/lib/screens/comments/widgets/comments_list.dart`
- `mobile/packages/divine_ui/lib/src/bottom_sheet/vine_bottom_sheet.dart`

## Desired behavior

- Opening the keyboard while composing a top-level comment, reply, or edit keeps the full composer visible above the keyboard.
- Reply mode remains usable on compact phone heights.
- The send button and keyboard-dismiss button remain tappable while the keyboard is open.
- Draft text is preserved if the keyboard is dismissed.
- Existing list interactions still dismiss the keyboard where intended.
- The fix does not regress other `VineBottomSheet` callers.

## Implementation notes

- Start from a fresh worktree branched from `origin/main`.
- Reproduce the issue with a widget test using simulated `viewInsets`.
- Prefer fixing at the comments call site first, not by changing global `VineBottomSheet` behavior.
- Use the existing `DraggableScrollableController` support in `VineBottomSheet.show`.
- Pass a controller through `showVideoPausingVineBottomSheet`.
- Expand the comments sheet to `maxChildSize` when the composer focus node gains focus or reply/edit mode requests focus.
- Keep the existing `_KeyboardAwareBottomInput` behavior intact.
- While the composer is focused, prevent dragging/snapping back to a size that can hide the composer, or otherwise prove with tests that dragging down cannot re-occlude the input.
- Do not introduce arbitrary `Future.delayed()` calls.

## Tests

Add or update widget tests that cover:

- focusing the top-level comment input with a simulated keyboard inset keeps the composer above the keyboard;
- activating reply mode focuses the composer and expands the sheet;
- reply composer controls remain visible/tappable above a simulated keyboard inset;
- keyboard dismissal remains available and does not clear draft text.

Run from `mobile/`:

- targeted comments widget tests
- `flutter analyze`

## Acceptance criteria

- The bug is reproduced by a failing test before the fix.
- The fixed comments sheet keeps the reply composer and send/dismiss controls above the keyboard.
- The fix is scoped to comments unless a broader shared-component change is proven necessary.
- No skipped tests, no TODO debt, no unrelated refactors.

### PR DESCRIPTION
**Pull request by NotThatKindOfDrLiz**
## What changed

- Threads a DraggableScrollableController through the comments bottom sheet call site.
- Expands the comments sheet to its max size when the composer gains focus, including reply/edit activation paths.
- Re-expands while the composer remains focused if the sheet is pulled below the safe comments snap size.
- Extracts comments sheet sizing values into named constants.
- Adds widget coverage for focus-driven expansion, focused-sheet reclamping, keyboard-dismiss availability, and draft retention.

## Why

The keyboard inset padding was applied inside the sheet's fixed fractional height. At smaller sheet sizes, the fixed chrome plus composer plus keyboard-height padding could exceed the visible sheet region and push the reply composer controls under the keyboard.

Closes #5939.

## Validation

- flutter analyze
- flutter test test/screens/comments/comments_screen_test.dart
- pre-push hook analyzer and changed-file tests passed

### CODE REVIEW

**Comment by NotThatKindOfDrLiz** [comment]
Updated this PR to address the review finding about the keyboard expansion tests exercising a copied implementation instead of production code.

What changed:
- Exposed the real comments composer widget as `MainCommentInput` with `@visibleForTesting`, matching the existing public-by-test pattern used by `OutboxBridges`.
- Removed the hand-copied `_MainCommentInputTest` widget and now pump the real production input in `comments_screen_test.dart`.
- Added the needed `FeatureFlag.videoReplies` override because the real widget reads that provider.
- Updated the reply-indicator assertion to match the production generated-name fallback instead of the old hardcoded `TestUser` test double behavior.

Local verification:
- `flutter analyze`
- `flutter test test/screens/comments/comments_screen_test.dart`
- pre-push hook analyzer + changed-file test

**Review by hm21 — APPROVED** [review]
LGTM. Pushed a small re-entrancy guard on top so the focus-driven expand runs as a single animation instead of restarting each frame.

**Review by dcadenas — APPROVED** [review]
✅ replies stay above the keyboard

**Review by realmeylisdev — COMMENTED** [review]
Formal review, verified against the Flutter 3.44 SDK. The focus-driven expand is the right mechanism and hm21's re-entrancy guard correctly fixes the real per-frame `animateTo` leak. Non-blocking: no Nostr / backend / protocol surface is touched, and the primary tap -> type -> send flow is unaffected. Three items inline: one correctness edge case (the `_isExpanding` latch can strand and silently disable the re-clamp for a sheet session) plus two test-coverage gaps.

File: mobile/lib/screens/comments/comments_screen.dart [hunk 1]
```diff
@@ -569,32 +579,76 @@ class _CommentsScreenBody extends StatelessWidget {
...
+    }
+
+    _isExpanding = true;
+    controller
+        .animateTo(
+          _CommentsSheetSizing.max,
+          duration: _CommentsSheetSizing.focusExpandDuration,
+          curve: Curves.easeOut,
+        )
+        .whenComplete(() => _isExpanding = false);
```
Comment (realmeylisdev) [inline mobile/lib/screens/comments/comments_screen.dart]: **Stranded latch can disable the re-clamp for a sheet session.**

`_isExpanding` is reset only here, via `animateTo(...).whenComplete(...)`. In Flutter 3.44 `DraggableScrollableController.animateTo` awaits an internal `AnimationController`'s `TickerFuture`; when that animation is *interrupted* (user drag -> `extent.addPixelDelta` -> `_cancelActivity()` -> `animationController.stop(canceled: true)` -> `TickerFuture._cancel`) only the `.orCancel` secondary completer resolves. The primary completer never completes, so `.whenComplete` never runs and `_isExpanding` stays `true` for the widget's lifetime. Since `_expandSheetForKeyboard` is the sole reset site and early-returns while `true`, both re-clamp paths (`_handleFocusChanged`, `_handleSheetSizeChanged`) then stop working.

Reachability is wider than the initial-focus window: while focused, dragging the sheet down through 0.7 fires a mid-drag re-clamp `animateTo` that the next drag delta cancels, stranding the latch. And `snapSizes: [0.7, 0.93]` with `minChildSize: 0.5` implies the snap set `[0.5, 0.7, 0.93]` (`DraggableScrollableSheet` prepends `minChildSize`), so the sheet can then rest at 0.5 with the composer back behind the keyboard, the exact condition #5939 fixes, until the sheet is closed and reopened.

Verified empirically: `animateTo(...).whenComplete(cb)` runs `cb` on normal completion but not when interrupted by a drag/`jumpTo`. Suggest resetting the latch off a reliable signal (clear it in the size listener when max is reached or an interruption is observed, or via the `.orCancel` error path) rather than the completion future. This also means the commit message's "resets on completion (including user-drag cancel)" doesn't hold for the cancel case.

File: mobile/test/screens/comments/comments_screen_test.dart [hunk 2]
```diff
@@ -475,6 +514,110 @@ void main() {
...
+            listState: listState,
+          ),
+        );
+        await tester.pump();
+
+        await tester.tap(find.byType(TextField).first);
+        await tester.pumpAndSettle();
+        expect(draggableController.size, closeTo(0.93, 0.001));
+
+        draggableController.jumpTo(0.5);
```
Comment (realmeylisdev) [inline mobile/test/screens/comments/comments_screen_test.dart:588]: **No coverage for the interrupted-expand path.**

This reaches the re-clamp from a fully-settled state: `pumpAndSettle()` lets the initial expand finish (so `_isExpanding` is already `false`) before `jumpTo(0.5)`, and `jumpTo` doesn't interrupt a running animation. So it can't catch the stranded-latch case: a test that starts a real drag *mid*-expand (`tester.drag` -> `addPixelDelta`) and then asserts a later sub-0.7 drop still re-clamps would fail against the current code. Worth adding, since the commit message claims this path is handled.

File: mobile/test/screens/comments/comments_screen_test.dart [hunk 3]
```diff
@@ -475,6 +514,110 @@ void main() {
...
+          await tester.pump();
+
+          expect(draggableController.size, 0.7);
+
+          await tester.tap(find.byType(TextField).first);
+          await tester.pumpAndSettle();
+
+          expect(draggableController.size, closeTo(0.93, 0.001));
+          expect(
+            find.bySemanticsIdentifier('send_comment_button'),
```
Comment (realmeylisdev) [inline mobile/test/screens/comments/comments_screen_test.dart]: **Asserts presence, not above-keyboard geometry (nit).**

`find.bySemanticsIdentifier('send_comment_button')` matches any laid-out node regardless of its y-position, so a composer painted *behind* the 300px `viewInsets` would still pass. The test topology (`_CommentsScreenTestContent` is a bare `Column`, not the real `VineBottomSheet`) also never consumes `viewInsets`, so the keyboard-avoidance padding isn't exercised. The meaningful guard here is the `draggableController.size == 0.93` assertion; the button-presence checks don't prove "above the keyboard".

File: mobile/lib/screens/comments/comments_screen.dart (hunk 1 above)
Comment (dcadenas) [inline mobile/lib/screens/comments/comments_screen.dart]: Clear `_isExpanding` on both animation completion and cancellation, then recheck the focused sheet size before returning control.
This reset currently depends on `whenComplete`, but Flutter can cancel the `DraggableScrollableController.animateTo` ticker when the user drags the sheet, leaving the latch set.
Once that happens, `_handleSheetSizeChanged` keeps calling `_expandSheetForKeyboard`, but the early return blocks the reclamp and the focused composer can stay below `_CommentsSheetSizing.initial`.
Add a widget test that focuses the composer, interrupts the in-flight expansion with real sheet input, and verifies the sheet can reclamp while focus remains.

**Review by dcadenas — CHANGES_REQUESTED** [review]
The expansion guard needs to be cancellation-safe before this can merge. As written, an interrupted sheet animation can leave the reply composer without the reclamp behavior this PR is meant to preserve.

File: mobile/lib/screens/comments/comments_screen.dart (hunk 1 above)
Comment (hm21) [reply mobile/lib/screens/comments/comments_screen.dart]: Fixed in 6c4f14fd06. `_isExpanding` is now released by a duration-matched `Timer` instead of the `animateTo` future — it fires on completion *or* drag-cancellation, so the latch can't strand. On release it re-checks the focused sheet size via `_handleSheetSizeChanged`, so an interrupted expand still re-clamps above the keyboard. Added a test that interrupts the in-flight expand with real sheet input and asserts the re-clamp while focus is retained.

Comment (hm21) [reply mobile/lib/screens/comments/comments_screen.dart]: Fixed in 6c4f14fd06, exactly as you diagnosed. The latch no longer hangs off the `animateTo` future; a `focusExpandDuration` `Timer` releases it (and re-checks the sheet size) regardless of whether the animation completed or a drag cancelled it. Timer is cancelled in `dispose`.

File: mobile/test/screens/comments/comments_screen_test.dart (hunk 2 above)
Comment (hm21) [reply mobile/test/screens/comments/comments_screen_test.dart:588]: Added in 6c4f14fd06: `re-clamps after an interrupted focus expand` starts the expand, interrupts it mid-flight with `jumpTo(0.5)`, elapses past the 250ms window, and asserts the sheet returns to max with focus retained. It goes red against the old whenComplete-based latch.

File: mobile/test/screens/comments/comments_screen_test.dart (hunk 3 above)
Comment (hm21) [reply mobile/test/screens/comments/comments_screen_test.dart]: Addressed in 6c4f14fd06 — dropped the misleading `viewInsets` setup (the bare-`Column` topology never consumed it) and the `send_comment_button` presence assertion. The `size == 0.93` fraction is now the guard, with the focus-gated `hide_comment_keyboard_button` proving focus actually took effect.

**Comment by hm21** [comment]
@dcadenas Good catch on the cancellation edge — we pushed the original re-entrancy guard, so we followed through and helped get the review points resolved in 6c4f14fd06.

The latch is now released by a duration-matched `Timer` rather than the `animateTo` future (which never resolves on a drag-cancel), and it re-checks the focused sheet size on release, so an interrupted expand still re-clamps above the keyboard. Added a regression test that interrupts the in-flight expand with real sheet input and asserts the re-clamp with focus retained, and tightened the existing focus test onto the sheet-fraction guard per @realmeylisdev's note. Replies left on each thread. Unblocking.

**Review by dcadenas — APPROVED** [review]
⌨️ reply composer stays above the keyboard

File: mobile/lib/screens/comments/comments_screen.dart [hunk 4]
```diff
@@ -301,12 +305,21 @@ abstract final class CommentsScreen {
             );
           },
         )
-        .whenComplete(
-          surfaceTelemetry.completeDismissed,
-        );
+        .whenComplete(() {
+          surfaceTelemetry.completeDismissed();
+          draggableController.dispose();
```
Comment (realmeylisdev) [inline mobile/lib/screens/comments/comments_screen.dart]: Non-blocking (debug/profile-only): disposing `draggableController` here can throw `A DraggableScrollableController was used after being disposed.`

`showModalBottomSheet`'s future resolves at pop-*start* (`Route.didComplete` completes `popped` before the exit transition), so this `whenComplete` runs while the sheet is still mounted and animating out. A focus-driven `_expandSheetForKeyboard` arms a 250ms `animateTo`, and `DraggableScrollableController` has no `dispose()` override — so `ChangeNotifier.dispose()` neither detaches nor stops that animation (both happen only in `_detach`, at sheet unmount). If the sheet is dismissed via a non-drag vector (scrim tap / system back / programmatic pop) within the animation window, the still-ticking `animateTo` keeps calling `notifyListeners` on the just-disposed controller and the `debugAssertNotDisposed` assert fires. Drag-dismiss is safe — the drag's `startActivity` cancels the animation first.

Reproduced against Flutter 3.44 with the exact pattern (`showModalBottomSheet` + `whenComplete` dispose + in-flight `animateTo` + non-drag pop mid-animation): throws on each exit frame. Asserts are stripped in release, so there's no shipped-user crash — but it's a real red-screen in debug and can surface in dev / integration runs; the production teardown path (whenComplete-dispose during an expand) isn't unit-tested.

Suggestion: tie disposal to the sheet subtree's own teardown (a widget whose `dispose()` runs at pop-end, after `_detach` has stopped the animation) rather than the route future's `whenComplete`.

**Review by realmeylisdev — COMMENTED** [review]
Formal re-review of the final 4-commit state (verified against Flutter 3.44).

The keyboard fix is sound and complete. Focus-driven expand to `maxChildSize` with a timer-released re-entrancy latch is correct, and the reclamp is genuinely reachable in production — `UnfocusOnSheetDismiss` only unfocuses on route reverse, so a partial focused drag keeps focus and gets clamped back up. All three items from the prior round are correctly resolved by 6c4f14fd0: the strand-latch (now a duration-matched `Timer`, not the `animateTo` future), the missing interrupt test, and the presence-vs-geometry assertion.

Verified locally: 30/30 `comments_screen_test.dart` pass, `flutter analyze` clean on both changed files, and the production plumbing genuinely attaches the controller (`showVideoPausingVineBottomSheet` → `VineBottomSheet.show` → the real `DraggableScrollableSheet`). No Nostr / backend / protocol surface is touched.

One non-blocking, debug-only lifecycle nit inline — not a merge blocker.

File: mobile/lib/screens/comments/comments_screen.dart (hunk 4 above)
Comment (realmeylisdev) [reply mobile/lib/screens/comments/comments_screen.dart]: Addressed in fe64f7b54 (pushed to this branch). `disposeCommentsSheetController` cancels any in-flight expand via `jumpTo` (which fires the animateTo's `onCanceled` → `AnimationController.stop`) before disposing, so nothing ticks on the disposed controller as the sheet animates out. Added a regression test that interrupts an in-flight expand mid-flight, disposes through the helper, and asserts no exception — it goes red against the previous plain-`dispose()` path (throws "used after disposed") and green with the fix. `flutter analyze` + all 31 `comments_screen_test.dart` cases pass.

**Review by realmeylisdev — APPROVED** [review]
Reviewed the full diff plus the sheet-controller lifecycle end to end. The keyboard-avoidance mechanism is sound: single `DraggableScrollableController` threaded through the bottom sheet and `MainCommentInput`, expand latch released on a duration-matched timer (correct, since the animateTo future never completes on drag-cancel), and a focus-guarded reclamp on size change.

The one issue from my prior pass — disposing the controller in `.whenComplete` while an in-flight `animateTo` was still ticking, tripping a debug-only "used after disposed" assert — is resolved in fe64f7b54, which cancels the expand via jumpTo before dispose and adds a regression test that goes red against a plain dispose.

Tests and analyze are green. LGTM.

**Review by rabble — APPROVED** [review]
**Verdict: Approve with nits** ✅

No correctness bugs found in the load-bearing logic. The `_isExpanding` latch being released on a duration-matched `Timer` rather than the `animateTo` future is exactly right — that future never resolves when a drag cancels the animation, and the "re-clamps after an interrupted focus expand" test pins the stranding scenario that would otherwise regress. `Timer` with a documented reason is the sanctioned pattern (not `Future.delayed`), and the disposal helper cancelling the in-flight expand before `dispose` is covered by a test that would fail against a plain `controller.dispose()`.

Nits (none blocking):
1. `comments_screen.dart:653-662` — the focused re-clamp floor is `>= _CommentsSheetSizing.initial` (0.7), so a focused sheet may rest at exactly 0.7. On a short device with a tall keyboard the composer can still be tight (visible area ≈ sheetFraction − keyboardFraction). Not a regression — the PR only ever raises focused sheets — but flagging so "0.7 is safe" stays a conscious assumption rather than an inherited one.
2. `comments_screen.dart:315-335` + `616-695` — vanishingly small residual window: a route removed *without* a pop doesn't drop focus until `UnfocusOnSheetDismiss.dispose`, so `whenComplete` could dispose the controller while the 250ms `_expandWindowTimer` is pending. In practice the element unmounts within a frame and the state's `dispose` cancels the timer first; every normal dismiss path drops focus synchronously. Not blocking.

Tests assert the observable sheet fraction reaching 0.93 (not just widget presence) and cover focus-expand, reclamp-below-safe, interrupted-expand re-clamp, draft retention, and disposal.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
